### ISSUE TITLE
Built PS Move API and OpenCV successfully on Win 64bit, not working (details inside).

### ISSUE THREAD

**Issue by ChairGraveyard** [thp/psmoveapi#116]
Hello all!

First of all, thank you Thomas for the great library :D

I've recently (about a week ago) started working on a Move plugin for Unreal Engine 4 (which I'm using to build a VR game). 

Unfortunately, UE4 does not have support for 32bit editor builds, or I'd probably just use that to test.

So that put me down the path of trying to recompile the API for 64bit. 

I got that working per the directions in the Windows readme, with MinGW-w64. OpenCV built successfull, and the PS Move API also.

The problem occurs when any data is requested from the camera in OpenCV (e.g., psmove_tracker_new). It never returns a valid image, so tracker_new will simply hang (due to the while (!image) loop). 

In the test_tracking program with the debug capture turned on, the capture window appears but is just grey, and never gets any data.

It's *only* the tracking library I'm having a problem with - the normal portion of the APi works fine with the 64bit built DLLs, and I'm able to connect to the controller and get button presses data back properly.

I built OpenCV using CMake and MinGW-w64, and the same for PS Move API.

I've tried it using the static/local OpenCV libs, but I can't seem to build my own versions of those via CMake (it outputs the .a files, but never the .libs), and when I try and compile the PS Move API against the precompiled static libs, it fails to find the OpenCV functions in the linker step.

I'd really love to get this working, as I'm so close to having it all hooked up to send position and rotation into UE4. 

I'm not so good with these tooks (CMake/MinGW etc.), but I can make do and get them compiled. This issue though is really stumping me.

If anyone else can provide a precompiled 64bit tracking DLL that would be awesome too.

I'll be happy to give any additional details.

One thing I haven't tried, is building OpenCV with Visual Studio after CMake-ing the solution files and stuff for that. 

Any help is appreciated!

**Comment by cboulay** [comment]
Recently, I've been working on getting this working with the PS3EYEDriver in Windows 8.1 x64. I can build libusb-1.0, I can build OpenCV. I modified the CMakeLists.txt and the options seem to be set correctly, but the build fails on ps3eye.cpp.
Upon further investigation, I can build the PS3EYEDriver from [its original repo](https://github.com/inspirit/PS3EYEDriver) without too much trouble (a few minor syntactical changes), 
but I am unable to build the [thp's psmoveapi branch of PS3EYEDriver](https://github.com/thp/PS3EYEDriver/tree/psmoveapi) even outside the psmoveapi directory. The two branches [are quite different](https://github.com/thp/PS3EYEDriver/compare/psmoveapi).

I will start working through the psmoveapi branch of the PS3EYEDriver to see what's wrong, but I also expect that'll break the OS X build. Further, I'm not great at C++. Any hacks I do are sure not to make it back upstream.

If there's anyone out there that may have made progress on a psmoveapi on Windows x64-compatible version of the PS3EYEDriver then please chime in.

**Comment by cboulay** [comment]
I was able to build, I have to switch campuses to test now.
I had to add `add_definitions(-std=c++11)` to CMakeLists.txt
Also, ps3eye.cpp was generating the following error:
```
psmoveapi\external\PS3EYEDriver\src\ps3eye.cpp:784:56: error: taking address of temporary array
nanosleep((struct timespec[]){{0, 100000000}}, NULL);
```
(Same thing a few lines below)
So I added `|| defined _WIN64` to the line above `#include <windows.h>`
Add above the offending functions I changed `#ifdef _MSC_VER` to `if defined(_MSC_VER) || defined(_WIN64)`

I'll post back here after I test.

**Comment by cboulay** [comment]
Success! I had to install the camera as a libusbK device using [Zadig](http://zadig.akeo.ie/). And reboot. And unplug-replug.

The tracker works in windows 64-bit!

**Comment by nitsch** [comment]
> Success! I had to install the camera as a libusbK device using
> [Zadig](http://zadig.akeo.ie/). And reboot. And unplug-replug.
> 
> The tracker works in windows 64-bit!

Very nice! Care to prepare a pull request that contains the necessary
code changes and descriptions?

BTW, since I am currently going through the build on Windows 7, fixing
any problems I encounter for a possible new release, I do have a
question: Which version of OpenCV and MinGW exactly are you using? I
assume you built it yourself and then built the PS Move API with
PSMOVE_USE_LOCAL_OPENCV enabled. Correct?

I just tried that on Windows 7, using OpenCV 2.4.9 and had to fix some
library paths in PS Move API's CMakeLists.txt for resolving linker
errors when building libpsmoveapi_tracker.dll. Did you encounter similar
problems that need fixing?

**Comment by cboulay** [comment]
I also have a Macbook Pro so I'll test the regular build process there to make sure I didn't break anything, then I'll do a pull request.

I've never actually done a pull request and I think it's going to be complicated because I had to modify the submodule. I'll figure it out.

I followed the build-osx-snapshot pretty closely, including git checkout --force 2.4.2 on OpenCV. I guess this is probably not necessary but I had to change the name of the libraries in CMakeLists to account for the version number (249 -> 242). I expect it would also work with 250.

I think the rest of your questions are answered in my README.Win64

```

       Building and running the PS Move API on Windows 8.1 64-bit
       ==========================================================

Requirements
------------

- MinGW (tdm64-gcc-4.9.?.exe)
	http://tdm-gcc.tdragon.net/download
- CMake
	http://www.cmake.org/cmake/resources/software.html
- Git
	https://msysgit.github.io/
	Be sure to add git to path for usage in Windows Command Prompt
	
(Note: Most of the following should be done in the mingw command prompt)

1. Get you clone of the psmoveapi
	git clone git://github.com/thp/psmoveapi.git
	cd psmoveapi
	We will call this directory $PSMOVEAPI_CHECKOUT

2. Init and update the submodules
	git submodule init
	git submodule update
	
3. libusb-1.0
	http://libusb.info/
	Downloads -> Latest Windows Binaries
	Unpack into $PSMOVEAPI_CHECKOUT/external/libusb-1.0
	
4. Download and build OpenCV
	cd $PSMOVEAPI_CHECKOUT
	git clone git://github.com/Itseez/opencv.git
	cd opencv
	rm -rf build
	mkdir -p build
	cd build
	git checkout --force 2.4.4
	run cmake-gui.
		Set the source directory to $PSMOVEAPI_CHECKOUT/opencv
		Set the build directory to $PSMOVEAPI_CHECKOUT/opencv/build
		Configure for MinGW Makefiles
		Unset the following options:
			BUILD_SHARED_LIBS, BUILD_PERF_TESTS, BUILD_TESTS, BUILD_DOCS, BUILD_opencv_flann
			BUILD_opencv_features2d, BUILD_opencv_objdetect, BUILD_opencv_photo
			BUILD_opencv_ts, BUILD_opencv_ml, BUILD_opencv_video, BUILD_opencv_java
			WITH_OPENEXR, WITH_FFMPEG, WITH_JASPER, WITH_TIFF
		Set CMAKE_INSTALL_PREFIX to $PSMOVEAPI_CHECKOUT/opencv/build/install
		Configure again
		Generate
	mingw32-make -j4
	mingw32-make install
cd ..\..


5. prepare a new build with cmake for the psmoveapi
	mkdir build
	cd build
	Run cmake-gui.
		Set $PSMOVEAPI_CHECKOUT as the source
		Set $PSMOVEAPI_CHECKOUT/build as the build directory
		Configure for MinGW Makefiles
		Uncheck BUILD_PYTHON_BINDINGS
		Check PSMOVE_USE_PS3EYE_DRIVER and PSMOVE_USE_LOCAL_OPENCV
		Configure again
		Check PSMOVE_USE_LOCAL_LIBUSB

6. finally build
    mingw32-make -j4

7. Install the camera libusb interface.
    Download and run Zadig: http://zadig.akeo.ie/
    Set the USB Camera's Interface 0 to use libusbK driver. If Interface 1 does not show up then Windows may have detected it properly (did on one of my machines, but not the other).
    Reboot.
    You may have to cycle the USB plug.

8. Good luck pairing over bluetooth.
```

**Comment by cboulay** [comment]
I tried testing in OSX and I ran into some problems using the build script. My internet connection is too flaky to reliably download the > 400 MB OpenCV repo. So I added `--branch 2.2.4 --depth 1` to the git checkout line and the whole build script completed, but build/test_tracker tracking was broken. The controller got progressively dimmer until it was barely on, and the camera view had a purple box presumably searching the scene erratically for the move controller. I have no idea if this is because my OpenCV checkout scheme was wrong, if there's a known problem using the PS3Eye and driver in Yosemite, or something else.

If I simply download an archive of latest OpenCV and build it with the correct directory structure for PSMOVE_USE_LOCAL_OPENCV then I cannot build psmoveapi. Clearly it is not compatible with the latest OpenCV.

So, as you can see, I have no idea if what I'm doing breaks the OSX build because I cannot test it properly.

(More in next comment).

**Comment by cboulay** [comment]
Even if I make the probably-true assumption that my changes don't break anything else, I can't seem to wrap my head around making a pull request when there are simultaneous changes within a submodule.

Can someone explain to me the proper workflow for that? I've been googling for a while but couldn't find what I want. I'll do a pull request on JUST PS3EYEDriver for now. That I can figure out.

**Comment by nitsch** [comment]
> - MinGW (tdm64-gcc-4.9.?.exe)
> 	http://tdm-gcc.tdragon.net/download

Any special reason for choosing this flavor of MinGW instead of the
MinGW-w64 version which is mentioned in README.win32? Could you try with
that version as well? I know it is working on Windows <= 7, but I cannot
test on Windows 8.


> 5. prepare a new build with cmake for the psmoveapi
> 	mkdir build
> 	cd build
> 	Run cmake-gui.
> 		Set $PSMOVEAPI_CHECKOUT as the source
> 		Set $PSMOVEAPI_CHECKOUT/build as the build directory
> 		Configure for MinGW Makefiles
> 		Uncheck BUILD_PYTHON_BINDINGS
> 		Check PSMOVE_USE_PS3EYE_DRIVER and PSMOVE_USE_LOCAL_OPENCV
> 		Configure again
> 		Check PSMOVE_USE_LOCAL_LIBUSB

So no CMake changes required to link the OpenCV libs when using OpenCV
2.4.4, except, of course, changing the suffixes from 249 to 244 in
CMakeLists.txt. Is that correct?

**Comment by nitsch** [comment]
> I tried testing in OSX and I ran into some problems using the build
> script. My internet connection is too flaky to reliably download the
> > 400 MB OpenCV repo. So I added `--branch 2.2.4 --depth 1` to the
> git checkout line and the whole build script completed, but
> build/test_tracker tracking was broken. The controller got
> progressively dimmer until it was barely on, and the camera view had
> a purple box presumably searching the scene erratically for the move
> controller. I have no idea if this is because my OpenCV checkout
> scheme was wrong, if there's a known problem using the PS3Eye and
> driver in Yosemite, or something else.

I do not believe that this behaviour has anything to do with the
checkout. The resulting code should be the same. It seems to be a
problem with the tracker, most likely the initial calibration. Maybe
caused by different camera settings.

Try building with the CMake option PSMOVE_USE_TRACKER_TRACE enabled.
This will make the calibration routine save debug images and some
accompanying .js text file (dunno where exactly this will end up on OS
X, but you will certainly figure that out). It is probably easier to
identify the problem when looking at these images.

**Comment by nitsch** [comment]
> Even if I make the probably-true assumption that my changes don't
> break anything else, I can't seem to wrap my head around making a
> pull request when there are simultaneous changes within a submodule.

Unfortunately, updates involving changes in submodules are a little
elaborate in this collaborative setup ...

You will have to issue a pull request for the changes in the submodule
first. Once that is merged by Thomas, you go back to psmoveapi and check
out the new version of the submodule from within the
external/PS3EYEDriver directory. Change back to the psmoveapi root
directory and commit this change as usual. You can then issue a pull
request for that change in psmoveapi.

**Comment by nitsch** [comment]
Regarding the pull request ... To make this clear: you have to issue
pull requests for the two different repositories (PS3EYEDriver and
psmoveapi).

**Comment by cboulay** [comment]
>Any special reason for choosing this flavor of MinGW instead of the MinGW-w64 version which is mentioned in README.win32? Could you try with that version as well? I know it is working on Windows <= 7, but I cannot test on Windows 8.

I just prefer the TDM distribution because I find it easier to maintain. I just tried with the version in README.win32 and everything worked as expected.

Also, I just built OpenCV 2.4.10 and put that in the expected OPENCV_INSTALL_DIR and I was able to build psmoveapi with PSMOVE_USE_LOCAL_OPENCV. Besides changing the lib suffixes below, I also had to change link_directories(${OPENCV_INSTALL_DIR}/x64/mingw/staticlib).

I cannot test if the tracker actually works on my MacbookPro in Bootcamp because libusb seems to fail when it discovers my apple trackpad. I think we need a whitelist or blacklist so that it simply doesn't attempt to use all the devices, but that's a separate issue.

I made a [pull request](https://github.com/thp/PS3EYEDriver/pull/1) for my PS3EYEDriver changes then I'll make a pull request for the psmoveapi changes once that is merged.

**Comment by nitsch** [comment]
> Also, I just built OpenCV 2.4.10 and put that in the expected
> OPENCV_INSTALL_DIR and I was able to build psmoveapi with
> PSMOVE_USE_LOCAL_OPENCV. Besides changing the lib suffixes below, I
> also had to change
> link_directories(${OPENCV_INSTALL_DIR}/x64/mingw/staticlib).

Yes, that is what seems to have changed in some release following OpenCV
2.4.4. Version 2.4.9 has the same output scheme, too.

**Comment by thp** [comment]
We now have regular automated test builds in CI, so I consider this fixed. There are even Visual Studio builds now.
